aws/data_source_aws_s3_bucket_objects.go

```diff
@@ -0,0 +1,183 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+const keyRequestPageSize = 1000
+
+func dataSourceAwsS3BucketObjects() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsS3BucketObjectsRead,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"prefix": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"delimiter": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"encoding_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"max_keys": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  1000,
+			},
+			"start_after": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"fetch_owner": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"keys": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"common_prefixes": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"owners": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func dataSourceAwsS3BucketObjectsRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket := d.Get("bucket").(string)
+	prefix := d.Get("prefix").(string)
+
+	input := &s3.HeadBucketInput{
+		Bucket: aws.String(bucket),
+	}
+
+	log.Printf("[DEBUG] Reading S3 bucket: %s", input)
+	_, err := conn.HeadBucket(input)
+
+	if err != nil {
+		return fmt.Errorf("Failed listing S3 bucket object keys: %s Bucket: %q", err, bucket)
+	}
+
+	d.SetId(fmt.Sprintf("%s_%s", bucket, prefix))
+
+	arn := arn.ARN{
+		Partition: meta.(*AWSClient).partition,
+		Service:   "s3",
+		Resource:  bucket,
+	}.String()
+	d.Set("arn", arn)
+
+	listInput := s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+	}
+
+	if prefix != "" {
+		listInput.Prefix = aws.String(prefix)
+	}
+
+	if s, ok := d.GetOk("delimiter"); ok {
+		listInput.Delimiter = aws.String(s.(string))
+	}
+
+	if s, ok := d.GetOk("encoding_type"); ok {
+		listInput.EncodingType = aws.String(s.(string))
+	}
+
+	// "listInput.MaxKeys" refers to max keys returned in a single request
+	// (i.e., page size), not the total number of keys returned if you page
+	// through the results. "maxKeys" does refer to total keys returned.
+	maxKeys := -1
```

Comment:
Since the schema attribute has a `Default`, this can be always retrieved via `d.Get()`:

```suggestion
	maxKeys := d.Get("max_keys").(int)
```